### ISSUE TITLE
Question: memory allocation inside wasm module

### ISSUE THREAD

**Issue by skrphv** [wasm3/wasm3#264]
Every time I trying to use something complex in wasm sources I got `"WASM3 error: restricted opcode"`.

For example `std:string` in cpp:
```cpp
#include <string>
extern "C" {
    #define WASM_EXPORT __attribute__((used)) __attribute__((visibility ("default")))
   
    extern void exPrint(const char* str);
    
    WASM_EXPORT
    void printHello() {
        std::string helloString {"Hello From Wasm!"};
        exPrint(helloString.c_str());
    }
}
```
compiled with `emcc src.cpp -o src.wasm -s STANDALONE_WASM -s ERROR_ON_UNDEFINED_SYMBOLS=0 -O0 --no-entry`

Or `Array<>` in typescript:
```js
declare function exSendInt8Array(len : i32) : void;

export function getInt8Array() : void {
  let arr = new Array<i8>(100)
  for (let i:i8 = 0; i < arr.length; i++) {
    arr[i] = i;
  }
  exSendInt8Array(arr.length)
}
```
compiled with `asc src.ts --binaryFile src.wasm --optimize`

Is it a problem of compiling WASM?
Or I have to attach every single libraries `by hand`?

I've seen `m3_api_lib.c` where is `m3_LinkLibC` function. But do I have to create the similar for all libraries that is used in my WASM sourcec?

Looks like this problem connected whith allocating on a heap commands / instructions, cause I also have the same problem with `malloc`...

I will be grateful for any help

**Comment by vshymanskyy** [comment]
Duplicate of #258 ?

**Comment by skrphv** [comment]
Not exactly, but maybe the problem is the same. I do not understand =)

**Comment by skrphv** [comment]
@vshymanskyy nevertheless... Do heap allocation successfully works on wasm3?

**Comment by vshymanskyy** [comment]
Absolutely. Many WASI apps perform allocations from heap: https://github.com/wasm3/wasm3/tree/main/test/wasi

**Comment by skrphv** [comment]
@vshymanskyy so if want to use heap allocation I have to use WASI? (and compile wasm3 with `BUILD_WASI` == "simple"/"metawasi"/"uvwasi")?

**Comment by vshymanskyy** [comment]
Yes, this is probably the easiest way

**Comment by vshymanskyy** [comment]
Ok, I reopened the issue. I'll try to reproduce it.

**Comment by skrphv** [comment]
@vshymanskyy Thanks a lot! Sorry for bothering. 
For information: ....
- I have looked at and tryed test wasi examples: it works - no questions.
- Now I try manually run [test/wasi/simple/test.wasm](https://github.com/wasm3/wasm3/blob/main/test/wasi/simple/test.wasm) in `platform/cpp/`-sources

**Comment by skrphv** [comment]
1. Added to `platforms/cpp/CMakeLists.txt`:
```cmake
set(BUILD_WASI "uvwasi" CACHE STRING "WASI implementation")
set_property(CACHE BUILD_WASI PROPERTY STRINGS none simple uvwasi metawasi)

include(FetchContent)
FetchContent_Declare(
  uvwasi
  GIT_REPOSITORY https://github.com/vshymanskyy/uvwasi.git
  GIT_TAG b063d686848c32a26119513056874f051c74258a
)

FetchContent_GetProperties(uvwasi)
if(NOT uvwasi_POPULATED)
  FetchContent_Populate(uvwasi)
  include_directories("${uvwasi_SOURCE_DIR}/include")
  add_subdirectory(${uvwasi_SOURCE_DIR} ${uvwasi_BINARY_DIR} EXCLUDE_FROM_ALL)
endif()
set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Dd_m3HasUVWASI")
target_link_libraries(${target} PRIVATE uvwasi_a uv_a)
```
2. Loaded wasm file `test/wasi/simple/test.wasm`
3. Result
```
Loading file...
Parsing wasm module... 
Loading wasm module... 
WASM3 error: restricted opcode
```

=(

**Comment by skrphv** [comment]
Ah! Didn't call `m3_LinkWASI`-function....

**Comment by vshymanskyy** [comment]
Does it work now?

**Comment by skrphv** [comment]
Fighting with linking uvwasi...

**Comment by skrphv** [comment]
```
RUNNING TEST: test20
wasi
Loading file...
Parsing wasm module... 
Linking WASI...
Segmentation fault (core dumped)
```

**Comment by skrphv** [comment]
Hmmmmm.... Looks like I have to link any function to `M3Module` only after loading this module to a `M3Runtime`...   
But if I loading module to a runtume before linking WASI I get `WASM3 error: restricted opcode`....   
It seems I compiled something wrong....

**Comment by skrphv** [comment]
Yep, linking function have to become after loading module to a runtume: [source/m3_bind.c#L134](https://github.com/wasm3/wasm3/blob/main/source/m3_bind.c#L134)

**Comment by skrphv** [comment]
@vshymanskyy hi! Got some surprise results...   
I have rewritten my code with wasm3 bare c-functions (like you do in [platforms/app/main.c](https://github.com/wasm3/wasm3/blob/main/platforms/app/main.c)) (without c++ api [platforms/cpp/main.cpp](https://github.com/wasm3/wasm3/blob/main/platforms/cpp/main.cpp))  
and my simple wasm programs and `test/wasi/simple/test.wasm` works fine.  
So highly likely there are bugs in cpp wrappers.

**Comment by skrphv** [comment]
@vshymanskyy not so rosy now....    

Cases like:

- 
```js
export function getStringAsArrayBuffer() : ArrayBuffer {
    let s : string = "Assembly Script String as ArrayBuffer"
    let arr = String.UTF8.encode(s, true)
    return arr
}
```
- or 
```js
export function getInt8Array() : Array<i8> {
    let arr = new Array<i8>(123) //- Error: [trap] out of bounds memory access
    for (let i:i8 = 0; i < 112; i++) {
        arr[i] = i
    }
    return arr;
}
```
- or calling built-in function `__new` 

leads to `[trap] out of bounds memory access`...

**Comment by skrphv** [comment]
@vshymanskyy update....   
Fixed up `[trap] out of bounds memory access` by compiling *.ts files with compiler command: 

```bash
asc  ...  --runtime stub
```

that means: 
```
--runtime             Specifies the runtime variant to include in the program.

                       incremental  TLSF + incremental GC (default)
                       minimal      TLSF + lightweight GC invoked externally
                       stub         Minimal runtime stub (never frees)
                       ...          Path to a custom runtime implementation
```


Does wasm3 works fine with GC (garbage collector)?

**Comment by vshymanskyy** [comment]
Is the initial bug in C++ wrapper fixed with https://github.com/wasm3/wasm3/pull/272 ?

**Comment by skrphv** [comment]
Partially.
1. If I use C/C++ as a wasm source, then `new`,` malloc`, `std::string`, etc. will work fine. (And without WASI)
1. If I use AssemblyScript as a wasm source, I have to use the compiler-specific option `--runtime stub`, which means `Minimal runtime stub (never frees)`. I don’t understand yet whether wasm3 works with the garbage collector or not.

So the original problem seems to have been resolved.

**Comment by vshymanskyy** [comment]
probably it's caused by access to 0 offset (which C/C++ tries to avoid)

**Comment by vshymanskyy** [comment]
Another memory allocation - related question: https://github.com/wasm3/wasm3/issues/260
